### PR TITLE
[4.3] Upstream 4.3 telnyx carrier api issues

### PR DESCRIPTION
**Pull request by bradfordben**
Fixes to the Telnyx carrier module

The best effort configuration setting did not work, My guess is that this was written and then telnyx changed the api response and the response to no longer included the flag “any_best_effort” so the old code would never filter the best effort out. This improvement will now pass the best effort flag to telnyx and remove the need to filter the response.

Requests for TF numbers was failing as the request built and passed to Telnyx was incorrect. It passed the TF NAP as `prefix` in the body and it should be `nap` as there api doc states. I have also added the option to filer TF numbers by NXX.

Also found that the crossbar quantity option was always overwritten to 500 so every time you requested 15 numbers through CB it requested 500 from telnyx resulting in longer query times. The fix was to only set the default values if they were not set.

Kazoo5 PR https://github.com/2600hz/kazoo-core/pull/48

### CODE REVIEW

File: core/kazoo_number_manager/src/carriers/knm_telnyx.erl [hunk 1]
```diff
@@ -65,24 +68,32 @@ check_numbers(_Numbers) -> {'error', 'not_implemented'}.
...
     Country = knm_search:country(Options),
     Results = numbers('region', Quantity, Country, 'undefined'),
     {'ok', international_numbers(Results, Options)}.
 
+%%------------------------------------------------------------------------------
+%% @doc For a given Number, parse out the NXX value.
+%% Return `undefined' if the NXX value can not be parsed.
+%% @end
+%%------------------------------------------------------------------------------
+-spec get_nxx(kz_term:ne_binary()) -> kz_term:ne_binary() | 'undefined'.
```
Comment (jamesaimonetti) [inline core/kazoo_number_manager/src/carriers/knm_telnyx.erl:90]: api_ne_binary

File: core/kazoo_number_manager/src/carriers/knm_telnyx.erl [hunk 2]
```diff
@@ -65,24 +68,32 @@ check_numbers(_Numbers) -> {'error', 'not_implemented'}.
...
+get_nxx(Num) ->
+    case byte_size(Num) >= 2+3+3 of
```
Comment (jamesaimonetti) [inline core/kazoo_number_manager/src/carriers/knm_telnyx.erl:92]: Pulled out the ole TI-82 and computed `2+3+3 = 8`! Is there a readability thing necessitating keeping the addition? :)

Also I wonder if you can just do:

```erlang
get_nxx(<<_PlusOneAreaCode:5/binary, NXX:3/binary, _/binary>>) -> NXX;
get_nxx(_) -> 'undefined'.
```
This seems a bit more readable than guessing why `2+3+3` is important.

File: core/kazoo_number_manager/src/knm_telnyx_util.erl [hunk 3]
```diff
@@ -66,13 +62,10 @@ req(Method, Path) ->
...
 req('post', ["number_searches"], JObj) ->
-    case kz_json:get_value([<<"search_descriptor">>, <<"prefix">>], JObj) of
-        <<"800">> -> rep_fixture("telnyx_tollfree_search.json");
-        'undefined' ->
-            case kz_json:get_value([<<"search_descriptor">>, <<"country_iso">>], JObj) of
-                <<"GB">> -> rep_fixture("telnyx_international_search.json");
-                'undefined' -> rep_fixture("telnyx_npa_search.json")
-            end
+    case kz_json:get_integer_value([<<"search_type">>], JObj) of
+        1 -> rep_fixture("telnyx_npa_search.json");
```
Comment (jamesaimonetti) [inline core/kazoo_number_manager/src/knm_telnyx_util.erl:66]: Generally dislike `magic` numbers like these. Can you `-define(NICE_NAME, 1)` and use `?NICE_NAME` instead? I have no idea what 1, 2, and 3 mean in this context...

File: core/kazoo_number_manager/src/knm_telnyx_util.erl [hunk 4]
```diff
@@ -142,9 +135,7 @@ http_options() ->
 rep({'ok', 200=Code, _Headers, <<"{",_/binary>>=Response}) ->
     ?DEBUG_APPEND("Response:~n~p~n~p~n~s~n", [Code, _Headers, Response]),
 
-    Routines = [fun(JObj) -> maybe_remove_best_effort(?SHOULD_KEEP_BEST_EFFORT, JObj) end
-               ,fun(JObj) -> maybe_filter_rates(?SHOULD_FILTER_RATES, JObj) end
-               ],
+    Routines = [fun(JObj) -> maybe_filter_rates(?SHOULD_FILTER_RATES, JObj) end],
```
Comment (jamesaimonetti) [inline core/kazoo_number_manager/src/knm_telnyx_util.erl:138]: I like using named functions when possible. Could we make this be:

```erlang
Routines = [fun maybe_filter_rates/1],
...

maybe_filter_rates(JObj) -> maybe_filter_rates(JObj, ?SHOULD_FILTER_RATES).
```

Correcting arg order in `maybe_filter_rates/2`.

Thoughts?

**Review by jamesaimonetti — CHANGES_REQUESTED** [review]

File: core/kazoo_number_manager/src/knm_telnyx_util.erl (hunk 4 above)
Comment (bradfordben) [reply core/kazoo_number_manager/src/knm_telnyx_util.erl:138]: Yeah think that is a nicer way of doing it 👍

File: core/kazoo_number_manager/src/knm_telnyx_util.erl (hunk 3 above)
Comment (bradfordben) [reply core/kazoo_number_manager/src/knm_telnyx_util.erl:66]: Yeah definitely a cleaner way of doing it. Will make the change

**Comment by jamesaimonetti** [comment]
Merged

**Comment by bradfordben** [comment]
Thanks

**Pull request closed without merging**